### ISSUE TITLE
Disentangle "jax_enable_x64" and default dtype

### ISSUE THREAD

**Issue by francois-rozet** [google/jax#22688]
### Description

Providing a way to enable `float64` without changing the default dtype/precision. Or even better, adding the option to choose the default dtype. For example,

```python
jax.config.update('jax_enable_x64', True)
jax.config.update('jax_default_dtype', jnp.float32)
```

### Motivation

Some algorithms, such as fast Fourier transforms and conjugate gradient methods, are unstable at 32-bit precision. It is therefore recommended to apply them at 64-bit precision. If such algorithms are part of some bigger procedure, the entire procedure must run at 64-bit precision. This is very wasteful, especially if the procedure involves neural networks.

The current solution is to rewrite the procedure while explicitly requesting the `float32` dtype at **every** array creation. For large code bases, this would is very tedious and it prevents the use of most libraries.

### Feature request

Providing a way to enable `float64` without changing the default dtype/precision. A more general solution would be to add the option to choose the default dtype:

```python
jax.config.update('jax_enable_x64', True)
jax.config.update('jax_default_dtype', jnp.float32)
```

### Alternatives

An alternative would be to rewrite the procedure while explicitly requesting the dtype `float32` at every array creation. However, for large code bases, this would be very tedious and would also prevent the use of libraries.

**Comment by jakevdp** [comment]
We attempted this in the past (see the `jax_default_dtype_bits` configuration). It still exists, and still kind of works, but it's entirely undocumented and untested so I wouldn't rely on it. We abandoned the approach because it didn't seem worth the maintenance cost (doubling all our CI costs to check that default dtypes are respected).

`jax_enable_x64` is a really problematic API, but we haven't been able to come up with a good solution that doesn't significantly impact current users.

**Comment by francois-rozet** [comment]
Actually, why is `float64` not available by default? I am asking because an easy solution would be to enabl_ (not set as default) the 64-bit precision by default and remove that pesky `UserWarning` altogether.

**Comment by jakevdp** [comment]
> Actually, why is float64 not available by default?

I don't understand this question. Are you asking why `jax_enable_x64` is `false` by default? If so, the answer is one of JAX's original goals was fast computation on accelerators (GPU, TPU) which don't support 64-bit values. It was a decision made early, and defaults are hard to change because changing them breaks every user.

**Comment by francois-rozet** [comment]
No, my question is why is it prohibited to **use** float64 by default?

**Comment by jakevdp** [comment]
That's the question I answered

**Comment by francois-rozet** [comment]
My question is ambiguous, let me reformulate. Currently, `jax_enable_x64` is the flag that enables to use 64-bit types **and** sets 64-bit types as the default.

Assuming that the default remain 32-bit types, is there a reason to disable 64-bit types?

**Comment by jakevdp** [comment]
`jax_enable_x64` technically does nothing to change the default output of functions.

`jax_enable_x64=False` means that any time a 64-bit value would be produced, it is downcast to a 32-bit value.

Some APIs do produce 64-bit outputs by default, for example `jnp.float64(1)` produces a float64 value, and `jax_enable_x64=False` causes them to produce a 32-bit value instead. But it has nothing explicitly to do with default dtypes.

Does that make sense?

> Assuming that the default remain 32-bit types, is there a reason to disable 64-bit types?

Yes: there is very good reason to disable 64-bit types entirely. Any 64-bit value that enters into a computation on an accelerator leads to either an error (for hardware that does not support 64-bit) or a slow computation (in general for hardware that supports it). NumPy-style APIs often generate 64-bit types by default. For example, if you add an int32 and a uint32, you get an int64: and this happens silently. Similarly, Python scalars are 64-bit: so when you pass `1.0` to a function that converts it to an array, the result is a float64. This is important to realize: **64-bit values tend to sneak into programs, no matter how careful you are about your defaults**. It's the nature of NumPy and Python.

In the early days of JAX, this led to a lot of problems when trying to run neural network models on accelerators, and so the `jax_enable_x64` flag was added to make sure that 64-bit values never enter the computation. Was it a good design decision? Probably not in the long run, but it solved the problem of the moment, and allowed the demo to run correctly that week, and paved the way for what JAX has become today.

Can we change this decision now? Yes and no. The problem is that as soon as you relax the behavior of the X64 flag so that it allows 64-bit values, then those values tend to sneak in at unintended places. This breaks code for important stakeholders. It's *really* hard to make changes to these kinds of defaults without breaking people in subtle or not-so-subtle ways: I spent nearly a year attempting to do that a while back, and I ended up having to give up on the project due to its infeasibility.

Does that help clarify things?

**Comment by francois-rozet** [comment]
Ok, the problem is casting values from Python and NumPy to JAX. Then a solution based on contexts is probably better. Thank you for your answer! I would be happy to help @yashk2810.

**Comment by yashk2810** [comment]
I have a prototype here: https://github.com/google/jax/pull/21472 but it still has some problems I need to figure out.

**Comment by sebffischer** [comment]
@jakevdp would you mind sharing what you tried back then or what your insights were? was the thing holding you back backwards compatibility or was it maybe the jit/eager duality? to me it seems that one problem in jax is that literals are auto converted at the jit boundary, which forces to commit to a specific dtype without knowing whether the literal might e.g. be added to an f64 in the function body. (my question is a bit selfish as I am building an jax inspired and xla based array computing framework for R and I have stumbled upon the problem discussed here).

**Comment by jakevdp** [comment]
The main issue back then was that this change was pretty invasive – it required changing the default behavior for many APIs, which often resulted in downstream breakages. Since it was a global config, testing it effectively would essentially require doubling our CI footprint. And in the end, the cost/benefit analysis didn't favor this degree of churn and development cost for a feature that would have comparatively few users.
